src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+    /// as well as stored in the History Api.
+    pub fn to_string(&self) -> String {
+        let path = self.path_segments.join("/");
+        let mut path = format!("/{}", path); // add the leading '/'
+        if let Some(ref query) = self.query {
+            path = format!("{}?{}", path, query);
+        }
+        if let Some(ref fragment) = self.fragment {
+            path = format!("{}#{}", path, fragment)
+        }
+        path
+    }
+
+    /// Gets the path segment at the specified index.
+    pub fn get_segment_at_index<'a>(&'a self, index: usize) -> Option<&'a str> {
+        self.path_segments.get(index).map(String::as_str)
+    }
+}
+
+impl From<Url> for RouteInfo {
+    fn from(url: Url) -> RouteInfo {
+        RouteInfo {
+            path_segments: url
+                .path_segments()
+                .expect("The route should always start with '/', so this should never error.")
+                .map(str::to_string)
+                .collect::<Vec<String>>(),
+            query: url.query().map(str::to_string),
+            fragment: url.fragment().map(str::to_string)
+        }
+    }
+}
+
+
+impl RouteService {
+
+    /// Creates a new route service
+    pub fn new() -> RouteService {
+        RouteService {
+            history: window().history(),
+            location: window().location().unwrap(),
+            event_listener: None,
+            callback: None
+        }
+    }
+
+
+    /// Creates the callback used in the main routing logic.
+    ///
+    /// The callback takes a string, parses it into a url, and then uses the result of that
+    /// to create a message that the component will use to update itself with.
+    // TODO I would like to include this in register_router
+    pub fn create_routing_callback<COMP, CTX>(context: &mut Env<CTX, COMP>) -> Callback<RouteResult>
+        where
+            COMP: Component<CTX>,
+            COMP::Msg: From<RouteResult>,
+            CTX: 'static
+    {
+        return context.send_back(|route_result: RouteResult| {
+            println!("Callback path changed {:?}", route_result); // Remove me
+            COMP::Msg::from(route_result)
+        })
+    }
+
+    /// Will return the current route info based on the location API.
+    // TODO this should probably return a RouteResult and avoid expecting
+    pub fn get_current_route_info(&mut self) -> RouteInfo {
+        // If the location api errors, recover by redirecting to a valid address
+        let href = self.get_location().expect("Couldn't get href from location Api");
+        let url = Url::parse(&href).expect("The href returned from the location api should always be parsable.");
+        RouteInfo::from(url)
+    }
+
+    /// Registers the router.
+    /// There can only be one router.
+    /// The component in which it is set up will be the source from which routing logic emanates.
+    pub fn register_router(&mut self, callback: Callback<RouteResult>)
```

Comment:
nit: I chose to name this `initialize` since it fits with `yew::initialize()`.

Reply:
`initialize()` works, but I chose `register_router()` because this isn't called at app start.

Conceptually, someone could write an app that has a bunch of possible states where the component that performs routing isn't even instantiated until they perform some task. This scenario may be inadvisable, but Yew currently allows for it. Because of this, the router component could be instantiated and destroyed multiple times. 

I feel like there is a significant functional difference between `yew::initialize()` and how the router hooks into a component to justify a different name.

Reply:
If the `RoutingService` is attached to the `Model` by convention instead of the `Env` then we could accomplish _everything_ within a single `new()` in the `Model::create` method. Then we wouldn't have these annoying `Option` types.

It also allows people to call `set_state` (or whatever) in the `Model::view()` method, which allows for buttons, etc to respond directly. Considering that you should not call `Callback`s within `update` I think this is practically a _requirement_.

What do you think?